File: SmartFoodDeliveryApp/Services/AIClient.swift
import Foundation

@Observable
final class AIClient {
    var isProcessing: Bool = false
    var lastResponse: String = ""

    func processVoiceCommand(_ transcript: String, menuItems: [MenuItem]) async -> AIOrderResponse {
        isProcessing = true
        defer { isProcessing = false }

        try? await Task.sleep(for: .seconds(1.5))

        let lowered = transcript.lowercased()

        var suggestedItems: [MenuItem] = []

        for item in menuItems {
            if lowered.contains(item.name.lowercased()) ||
               lowered.contains(item.category.rawValue.lowercased()) {
                suggestedItems.append(item)
            }
        }

        if suggestedItems.isEmpty {
            if lowered.contains("burger") || lowered.contains("hungry") {
                suggestedItems = menuItems.filter { $0.category == .burgers }.prefix(2).map { $0 }
            } else if lowered.contains("healthy") || lowered.contains("light") || lowered.contains("salad") {
                suggestedItems = menuItems.filter { $0.tags.contains("Healthy") }.prefix(3).map { $0 }
            } else if lowered.contains("spicy") {
                suggestedItems = menuItems.filter { $0.tags.contains("Spicy") || $0.category == .sushi }.prefix(2).map { $0 }
            } else if lowered.contains("comfort") || lowered.contains("warm") {
                suggestedItems = menuItems.filter { $0.tags.contains("Comfort") || $0.category == .pasta }.prefix(2).map { $0 }
            } else {
                suggestedItems = Array(menuItems.sorted { $0.rating > $1.rating }.prefix(3))
            }
        }

        let itemNames = suggestedItems.map { $0.name }.joined(separator: ", ")
        let response = "Based on your request, I'd suggest: \(itemNames). Shall I add these to your cart?"

        lastResponse = response
        return AIOrderResponse(message: response, suggestedItems: suggestedItems)
    }

    func getMoodSuggestions(mood: Mood, menuItems: [MenuItem]) -> [MenuItem] {
        switch mood {
        case .comfort:
            return menuItems.filter { $0.category == .burgers || $0.category == .pasta || $0.tags.contains("Comfort") }
        case .light:
            return menuItems.filter { $0.calories < 500 || $0.category == .salads }
        case .energizing:
            return menuItems.filter { $0.tags.contains("Energizing") || $0.category == .bowls }
        case .spicy:
            return menuItems.filter { $0.tags.contains("Spicy") || $0.category == .sushi }
        case .healthy:
            return menuItems.filter { $0.tags.contains("Healthy") || $0.nutritionInfo.protein > 25 }
        case .indulgent:
            return menuItems.filter { $0.tags.contains("Indulgent") || $0.category == .desserts }
        }
    }

    func getBudgetSuggestions(maxBudget: Double, menuItems: [MenuItem]) -> [MenuItem] {
        menuItems.filter { $0.price <= maxBudget }.sorted { $0.rating > $1.rating }
    }

    func getNutritionSuggestions(filter: NutritionFilter, menuItems: [MenuItem]) -> [MenuItem] {
        switch filter {
        case .highProtein:
            return menuItems.filter { $0.nutritionInfo.protein >= 25 }.sorted { $0.nutritionInfo.protein > $1.nutritionInfo.protein }
        case .lowCalorie:
            return menuItems.filter { $0.calories <= 500 }.sorted { $0.calories < $1.calories }
        case .lowSugar:
            return menuItems.filter { $0.nutritionInfo.sugar <= 10 }.sorted { $0.nutritionInfo.sugar < $1.nutritionInfo.sugar }
        case .vegetarian:
            return menuItems.filter { !$0.ingredients.contains("Beef") && !$0.ingredients.contains("Chicken") && !$0.ingredients.contains("Shrimp") && !$0.ingredients.contains("Salmon") && !$0.ingredients.contains("Tuna") }
        }
    }
}

nonisolated struct AIOrderResponse: Sendable {
    let message: String
    let suggestedItems: [MenuItem]
}

nonisolated enum NutritionFilter: String, CaseIterable, Sendable {
    case highProtein = "High Protein"
    case lowCalorie = "Low Calorie"
    case lowSugar = "Low Sugar"
    case vegetarian = "Vegetarian"

    var icon: String {
        switch self {
        case .highProtein: "figure.strengthtraining.traditional"
        case .lowCalorie: "flame"
        case .lowSugar: "leaf"
        case .vegetarian: "carrot"
        }
    }
}

File: src/services/AIClient.ts
import type { MenuItem } from '../models/MenuItem';
import type { Mood } from '../models/Cart';

export type NutritionFilter =
  | 'highProtein'
  | 'lowCalorie'
  | 'lowSugar'
  | 'vegetarian';

export interface AIOrderResponse {
  message: string;
  suggestedItems: MenuItem[];
}

/**
 * Heuristic, local-only recommendation engine.
 * Mirrors the Swift `AIClient` behavior without using any LLM.
 */
export class AIClient {
  isProcessing = false;
  lastResponse = '';

  async processVoiceCommand(
    transcript: string,
    menuItems: MenuItem[],
  ): Promise<AIOrderResponse> {
    this.isProcessing = true;
    try {
      // Simulate a small thinking delay like the Swift version.
      await new Promise(resolve => setTimeout(resolve, 1500));

      const lowered = transcript.toLowerCase();
      let suggestedItems: MenuItem[] = [];

      for (const item of menuItems) {
        if (
          lowered.includes(item.name.toLowerCase()) ||
          lowered.includes(item.category.toLowerCase())
        ) {
          suggestedItems.push(item);
        }
      }

      if (suggestedItems.length === 0) {
        if (lowered.includes('burger') || lowered.includes('hungry')) {
          suggestedItems = menuItems
            .filter(m => m.category === 'burgers')
            .slice(0, 2);
        } else if (
          lowered.includes('healthy') ||
          lowered.includes('light') ||
          lowered.includes('salad')
        ) {
          suggestedItems = menuItems
            .filter(m => m.tags.includes('Healthy'))
            .slice(0, 3);
        } else if (lowered.includes('spicy')) {
          suggestedItems = menuItems
            .filter(
              m =>
                m.tags.includes('Spicy') || m.category === 'sushi',
            )
            .slice(0, 2);
        } else if (
          lowered.includes('comfort') ||
          lowered.includes('warm')
        ) {
          suggestedItems = menuItems
            .filter(
              m =>
                m.tags.includes('Comfort') ||
                m.category === 'pasta',
            )
            .slice(0, 2);
        } else {
          suggestedItems = [...menuItems]
            .sort((a, b) => b.rating - a.rating)
            .slice(0, 3);
        }
      }

      const itemNames = suggestedItems.map(i => i.name).join(', ');
      const message =
        suggestedItems.length > 0
          ? `Based on your request, I suggest: ${itemNames}. Shall I add these to your cart?`
          : `I did not find anything that matches your request.`;

      this.lastResponse = message;
      return { message, suggestedItems };
    } finally {
      this.isProcessing = false;
    }
  }

  getMoodSuggestions(mood: Mood, menuItems: MenuItem[]): MenuItem[] {
    switch (mood) {
      case 'comfort':
        return menuItems.filter(
          m =>
            m.category === 'burgers' ||
            m.category === 'pasta' ||
            m.tags.includes('Comfort'),
        );
      case 'light':
        return menuItems.filter(
          m => m.calories < 500 || m.category === 'salads',
        );
      case 'energizing':
        return menuItems.filter(
          m => m.tags.includes('Energizing') || m.category === 'bowls',
        );
      case 'spicy':
        return menuItems.filter(
          m =>
            m.tags.includes('Spicy') || m.category === 'sushi',
        );
      case 'healthy':
        return menuItems.filter(
          m =>
            m.tags.includes('Healthy') ||
            m.nutritionInfo.protein > 25,
        );
      case 'indulgent':
        return menuItems.filter(
          m =>
            m.tags.includes('Indulgent') ||
            m.category === 'desserts',
        );
      default:
        return [];
    }
  }

  getBudgetSuggestions(maxBudget: number, menuItems: MenuItem[]): MenuItem[] {
    return menuItems
      .filter(m => m.price <= maxBudget)
      .sort((a, b) => b.rating - a.rating);
  }

  getNutritionSuggestions(
    filter: NutritionFilter,
    menuItems: MenuItem[],
  ): MenuItem[] {
    switch (filter) {
      case 'highProtein':
        return menuItems
          .filter(m => m.nutritionInfo.protein >= 25)
          .sort(
            (a, b) =>
              b.nutritionInfo.protein - a.nutritionInfo.protein,
          );
      case 'lowCalorie':
        return menuItems
          .filter(m => m.calories <= 500)
          .sort((a, b) => a.calories - b.calories);
      case 'lowSugar':
        return menuItems
          .filter(m => m.nutritionInfo.sugar <= 10)
          .sort((a, b) => a.nutritionInfo.sugar - b.nutritionInfo.sugar);
      case 'vegetarian':
        return menuItems.filter(m => {
          const ingredients = m.ingredients;
          const animalProteins = [
            'beef',
            'chicken',
            'shrimp',
            'salmon',
            'tuna',
          ];
          return !ingredients.some(ing =>
            animalProteins.includes(ing.toLowerCase()),
          );
        });
      default:
        return [];
    }
  }
}

export const aiClient = new AIClient();

